### PR TITLE
Use IMDSv2 to access EC2 metadata

### PR DESCRIPTION
**Pull request by dsrosario**

### CODE REVIEW

**Comment by motobob** [comment]
please note #754 which was already merged.

File: src/erlcloud_aws.erl [hunk 1]
```diff
@@ -931,7 +931,9 @@ get_availability_zone() ->
 cache_instance_metadata_availability_zone() ->
     % it fine to use default here - no IAM is used, only for http client
     % one cannot use auto_config()/default_cfg() as it creates an infinite recursion.
-    case erlcloud_ec2_meta:get_instance_metadata("placement/availability-zone", #aws_config{}) of
+    Config = #aws_config{},
+    IMDSv2Token = maybe_imdsv2_session_token(Config),
```
Comment (motobob) [inline src/erlcloud_aws.erl:935]: are you saying this should ALWAYS be the new default behaviour?

Comment (dsrosario) [reply src/erlcloud_aws.erl:935]: Yes.. IMDSv2 should be the default behaviour due to security concerns. you can check https://www.uptycs.com/blog/you-should-be-using-aws-imdsv2-heres-why-and-how-to-do-it

Comment (dsrosario) [reply src/erlcloud_aws.erl:935]: Not requiring IMDSv2 is currently consider a high severity issue. In order to required IMDSv2 in EC2, all the software should be compliant.
Currently, from our test, we know that at least erlcloud_ddb2 will fail when trying to access dynamodb.

File: src/erlcloud_aws.erl [hunk 2]
```diff
@@ -974,13 +976,14 @@ timestamp_to_gregorian_seconds(Timestamp) ->
 get_credentials_from_metadata(Config) ->
     %% TODO this function should retry on errors getting credentials
     %% First get the list of roles
-    case erlcloud_ec2_meta:get_instance_metadata("iam/security-credentials/", Config) of
+    IMDSv2Token = maybe_imdsv2_session_token(Config),
+    case erlcloud_ec2_meta:get_instance_metadata("iam/security-credentials/", Config, IMDSv2Token) of
```
Comment (motobob) [inline src/erlcloud_aws.erl:980]: maybe I made a stylish mistake with previous PR and would prefer `Config` to always be the last parameter, but I guess it's late

File: src/erlcloud_aws.erl (hunk 1 above)
Comment (dsrosario) [reply src/erlcloud_aws.erl:935]: This is the official AWS documentation about the path to transition to IMDSv2

Comment (motobob) [reply src/erlcloud_aws.erl:935]: cool. let me read little more about it too. 
but why not ask just right here - do IAM roles attached to host need new permissions?

File: src/erlcloud_aws.erl [hunk 3]
```diff
@@ -1562,3 +1565,10 @@ get_env_for_role_credentials(Arn, ExtId) ->
 set_env_for_role_credentials(Arn, ExtId, Val) ->
     % application:set_env is undocumented in regards to type(Par) =/= atom()
     application:set_env(erlcloud, {role_credentials, Arn, ExtId}, Val).
+
+-spec maybe_imdsv2_session_token(aws_config()) -> binary() | undefined.
+maybe_imdsv2_session_token(Config) ->
+    case erlcloud_ec2_meta:generate_session_token(60, Config) of
```
Comment (motobob) [inline src/erlcloud_aws.erl:1571]: let's fix this in test and good to go.

Failure/Error: {error,undef,
                        [{erlcloud_ec2_meta,generate_session_token,
                             [60,
                              {aws_config,

**Comment by motobob** [comment]
thank you very much @dsrosario for contributing. bring more if you find anything!

**Comment by dsrosario** [comment]
Is it possible to release 3.7.2?

**Pull request merged**